src/erl/typespec.rs

```diff
@@ -0,0 +1,263 @@
+                    .clone()
+                    .unwrap_or("".to_string())
+                    .to_doc()
+                    .append(" :: ")
+                    .append(type_ast_name(&ast, None))
+            })
+            .intersperse(", ".to_doc()),
+    )
+    .surround("{", "}")
+}
+
+// Generates the spec for a regular Gleam function
+pub fn fun_spec(name: &str, args: &[TypedArg], return_type: &Type) -> Document {
+    "-spec "
+        .to_doc()
+        .append(atom(name.to_string()))
+        .append(spec_args(args))
+        .append(" -> ")
+        .append(typ_to_erl(return_type))
+        .append(".")
+}
+
+fn spec_args(args: &[TypedArg]) -> Document {
+    wrap_args(args.iter().map(|a| typ_to_erl(&*a.typ)))
+}
+
+fn typ_to_erl(typ: &Type) -> Document {
+    match typ {
+        Type::App { name, .. } => name.to_snake_case().to_doc(),
+        _ => "term".to_doc(),
+    }
+}
+
+// Generates the spec for an external Gleam function that wraps an erlang one
+pub fn external_fun_spec(name: &str, args: &[ExternalFnArg], return_type: &Type) -> Document {
+    // Find all parameters used in the return type - these are the only ones we can show in the
+    // args otherwise they are flagged as unused
+    let return_type_params = extract_params_from_type(return_type);
+    "-spec "
+        .to_doc()
+        .append(atom(name.to_string()))
+        .append(external_spec_args(args, &return_type_params))
+        .append(" -> ")
+        .append(typ_to_erl(return_type))
+        .append(".")
+}
+
+fn external_spec_args(args: &[ExternalFnArg], allowed_params: &[String]) -> Document {
+    wrap_args(
+        args.iter()
+            .map(|arg| type_ast_name(&arg.typ, Some(allowed_params))),
+    )
+}
+
+// Extracts the type names from a TypeAst - We optionally pass in a list of type parameters that we
+// are aware of.
+fn type_ast_name(ast: &TypeAst, allowed_params: Option<&[String]>) -> Document {
+    match ast {
+        TypeAst::Var { name, .. } =>
+        // If we provide a list of allowed type parameter names then we're in a context with
+        // limited possible type parameter names and if we come across one that isn't 'allowed'
+        // then we're better just outputting 'any' instead. This prevents unknown or unused
+        // type parameter warnings from erlang
+        {
+            allowed_params
+                .map(|slice| {
+                    if slice.contains(name) {
+                        name.to_camel_case()
+                    } else {
+                        "any".to_string()
+                    }
+                })
+                .unwrap_or(name.to_camel_case())
+                .to_doc()
+        }
+        TypeAst::Constructor { name, args, .. } => type_to_erl(name, &args, allowed_params),
+        _ => "term".to_doc(),
+    }
+}
+
+fn type_to_erl(name: &str, args: &[TypeAst], allowed_params: Option<&[String]>) -> Document {
+    match name {
+        "String" => "binary()".to_doc(),
```

Comment:
We're going to have to generate Erlang types from Gleam types as there's not enough information in the un-typed AST.

For example this code will cause the generation of incorrect Erlang

```rust
pub type Bool {
  Maybe
}

Reply:
I can see that's an issue. I'm not sure exactly what the answer is though.